lib/saml_idp/request.rb

```diff
@@ -150,10 +151,15 @@ def valid?
         errors.push(:no_response_url)
       end
 
-      unless service_provider? && valid_signature?
-        log "Signature is invalid in #{raw_xml}"
-        # TODO: We should get more specific errors
-        errors.push(:invalid_signature)
+      if service_provider?
+        begin
+          unless valid_signature?
+            log "Signature is invalid in #{raw_xml}"
+            errors.push(:invalid_signature)
+          end
+        rescue SamlIdp::XMLSecurity::SignedDocument::ValidationError => e
+          errors.push(validation_error_dictionary[e.message])
```

Comment:
1. do we need to have a default mesage here if the dictionary doesn't have a matching error?
2. the dictionary keys appear to be symbols but `e.message` looks like it will be strings

Reply:
1. i included all possible `ValidationError`s that could pop up (and some that actually i think are impossible for us to reach, but wanted to keep them there for the moment until we refactored the code) so it didn't feel strictly necessary. happy to add a default if you disagree![
2. i threw a `with_indifferent_access` in there but i can change it to the hash rockets if you'd prefer, i just think they're kind ugly!

Reply:
1. I didn't realize our own code threw those errors, if it's more limited that seems fine then. Maybe we just update each time we throw to have its own `error_code` property or something do we can do `e.error_code` instead of a reverse lookup
2. I missed the `with_indifferent_access` that seems fine